### PR TITLE
Add some extra functionality to improve debugging inside of docker

### PR DESCRIPTION
**Pull request by Bidski**
Adds address sanitisation and writes Asan log to disk to make it easier to read.
Adds `--asan` switch to `./b run` to control Asan writing to screen or writing to disk

Adds `--gdb` and `--valgrind` switches to `./b run` to facilitate running roles inside of gdb or valgrind.

Some extra information (to be added to NUBook as well) [Comparison Of Memory Tools](https://github.com/google/sanitizers/wiki/AddressSanitizerComparisonOfMemoryTools)

[Using ASan with GDB](https://github.com/google/sanitizers/wiki/AddressSanitizerAndDebugger)

[NUbook PR](https://github.com/NUbots/NUbook/pull/47)

### CODE REVIEW

**Comment by TrentHouliston** [comment]
Asan and valgrind do the same thing, but gdb does something a little different

File: tools/run.py [hunk 1]
```diff
@@ -11,18 +11,63 @@ def register(command):
...
 
+    # Get current environment
+    env = os.environ
+
+    # Add necessary ASAN environment variables
+    if use_asan:
+        env.update({"ASAN_OPTIONS": "log_path=/home/nubots/NUbots/asan.log"})
+
+    if use_gdb:
+        cmd = cmd + [
```
[the remark this answers is not in the document]
Comment (Bidski) [reply tools/run.py:50]: I thought i fixed that already ... 😞

**Comment by Bidski** [comment]
Asan also works a lot quicker than valgrind but probably gives less detailed information. I think you can also use asan with gdb

**Comment by ysims** [comment]
Ah yeah okay, but still the point stands with asan/valgrind.

**Comment by ysims** [comment]
Is it possible to detect if ASAN is set to ON in configure when running to stop valgrind happening and warn the user to turn off+rebuild, and executing whatever it does with `--asan` if it is on?

**Comment by Bidski** [comment]
> Is it possible to detect if ASAN is set to ON in configure when running to stop valgrind happening and warn the user to turn off+rebuild, and executing whatever it does with `--asan` if it is on?

So you want to remove the `--asan` option and replace it with a check to see if `USE_ASAN` is set?

**Pull request merged**